### PR TITLE
8258894: C2: Forbid GCM to move stores into loops

### PR DESCRIPTION
**Pull request by robcasloz**
Prevent GCM from placing memory-writing nodes (such as stores) into loops deeper than their home loop (determined by their control input). Such placements are invalid, as they cause memory definitions to interfere, and risk causing miscompilations. This change complements [JDK-8255763](https://bugs.openjdk.java.net/browse/JDK-8255763), which only addresses invalid placements in irreducible CFGs.

Add control input to stores in generated stubs to ensure that all memory-writing nodes have control inputs from which their home block can be derived.

Add a battery of simplified fuzzer test cases where, before this change, GCM moves stores into deeper loops.

Tested functionality on:
- hs-tier1-5 (windows-x64, linux-x64, linux-aarch64, and macosx-x64; release and debug mode)
- hs-tier6-9 (linux-x64, debug mode)

Tested performance on a set of standard benchmark suites (DaCapo, Renaissance, SPECjbb2015, SPECjbb2005, SPECjvm2008, ...) and on windows-x64, linux-x64, and macosx-x64. No regression was observed, which is expected since this change only affects the placement of memory-writing nodes (which tends to be quite constrained already), and only forbids placements which would be typically detrimental for performance anyway (moving such nodes into deeper loops).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8258894](https://bugs.openjdk.java.net/browse/JDK-8258894): C2: Forbid GCM to move stores into loops 

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2140/head:pull/2140`
`$ git checkout pull/2140`

### CODE REVIEW

**Comment by robcasloz** [comment]
/summary
Prevent GCM from placing memory-writing nodes (such as stores) into loops deeper
than their home loop (determined by their control input). Such placements are
invalid, as they cause memory definitions to interfere, and risk causing
miscompilations. This change complements JDK-8255763, which only addresses
invalid placements in irreducible CFGs.
    
Add control input to stores in generated stubs to ensure that all memory-writing
nodes have control inputs from which their home block can be derived.
    
Add a battery of simplified fuzzer test cases where, before this change, GCM
moves stores into deeper loops.

File: src/hotspot/share/opto/block.cpp [hunk 1]
```diff
@@ -1208,6 +1208,22 @@ void PhaseCFG::dump_headers() {
   }
 }
 
+void PhaseCFG::verify_memory_writer_placement(const Block* b, const Node* n) const {
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/block.cpp:1211]: You could wrap the entire method in `#ifdef ASSERT` and use `NOT_DEBUG_RETURN` in the .hpp (similar to  `verify_strip_mined`).

File: src/hotspot/share/opto/block.cpp [hunk 2]
```diff
@@ -1221,6 +1237,9 @@ void PhaseCFG::verify() const {
       if (j >= 1 && n->is_Mach() && n->as_Mach()->ideal_Opcode() == Op_CreateEx) {
         assert(j == 1 || block->get_node(j-1)->is_Phi(), "CreateEx must be first instruction in block");
       }
+      if (n->is_memory_writer()) {
+        verify_memory_writer_placement(block, n);
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/block.cpp:1241]: Other calls to this method are guarded by `#ifdef ASSERT`. Would it make sense to move the `n->is_memory_writer()` into the method, always call it and define it as `NOT_DEBUG_RETURN`?

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. Just to clarify, with JDK-8255763 in place, incorrect placement in the reducible CFG case is always prevented by the frequency estimation heuristic, right? I.e. without this patch, your newly added tests wouldn't fail? Would it make sense to run the test with `StressGCM` to trigger actual failures?

**Comment by robcasloz** [comment]
> Looks good to me.

Thanks for reviewing, Tobias!

> Just to clarify, with JDK-8255763 in place, incorrect placement in the reducible CFG case is always prevented by the frequency estimation heuristic, right?

Even for reducible CFGs there are cases (e.g. `testRegularReducible[2|3|4]()`) where GCM moves stores into loops deeper than their home loop, causing memory definitions to interfere _statically_: this is what I refer to as "invalid". However, in all these cases, the inner loop into which the store is moved has a trip count of one (otherwise its estimated frequency would be higher and the store would not be placed there), so there is no _dynamic_ interference and hence no possible miscompilation.

> I.e. without this patch, your newly added tests wouldn't fail?

Right, no assertion would fail (because C2 does not verify memory liveness after GCM), and no miscompilation would happen.  

> Would it make sense to run the test with `StressGCM` to trigger actual failures?

Yes, but triggering failures would require [JDK-8257146](https://bugs.openjdk.java.net/browse/JDK-8257146), which is blocked by this issue, to be integrated. I propose to add `StressGCM` to the tests later in JDK-8257146. Does that make sense?

File: src/hotspot/share/opto/block.cpp (hunk 1 above)
Comment (robcasloz) [reply src/hotspot/share/opto/block.cpp]: Done, thanks!

File: src/hotspot/share/opto/block.cpp (hunk 2 above)
Comment (robcasloz) [reply src/hotspot/share/opto/block.cpp]: Done, thanks!

**Comment by robcasloz** [comment]
I just applied the refactoring suggested by Tobias (commit d8ff541), re-tested on hs-tier1 (windows-x64, linux-x64, linux-aarch64, and macosx-x64; release and debug mode), and checked manually that the refactored verification code is executed in debug mode.

**Review by vnkozlov — APPROVED** [review]
Nice.

**Comment by TobiHartmann** [comment]
Thanks for the clarifications!

> Yes, but triggering failures would require JDK-8257146, which is blocked by this issue, to be integrated. I propose to add StressGCM to the tests later in JDK-8257146. Does that make sense?

Yes, that makes sense.

**Review by TobiHartmann — APPROVED** [review]

**Comment by robcasloz** [comment]
Thanks for reviewing, Tobias and Vladimir!

**Comment by robcasloz** [comment]
Please sponsor.

**Pull request closed without merging**